### ISSUE TITLE
Add linting to pytests

### ISSUE THREAD

**Issue by cosama** [lbl-anp/becquerel#125]
As @jccurtis pointed out, we should think about adding linting to the pytests. The issues raised by `pylint` can be questioned for their validity, but there is definitely plenty to improve:

```
+-----------------------------+------------+
|message id                   |occurrences |
+=============================+============+
|len-as-condition             |18          |
+-----------------------------+------------+
|fixme                        |13          |
+-----------------------------+------------+
|line-too-long                |11          |
+-----------------------------+------------+
|no-else-return               |9           |
+-----------------------------+------------+
|too-many-instance-attributes |6           |
+-----------------------------+------------+
|too-many-statements          |5           |
+-----------------------------+------------+
|too-many-branches            |5           |
+-----------------------------+------------+
|redefined-builtin            |3           |
+-----------------------------+------------+
|duplicate-code               |3           |
+-----------------------------+------------+
|protected-access             |2           |
+-----------------------------+------------+
|abstract-method              |2           |
+-----------------------------+------------+
|wrong-import-order           |1           |
+-----------------------------+------------+
|unused-variable              |1           |
+-----------------------------+------------+
|too-many-public-methods      |1           |
+-----------------------------+------------+
|too-many-locals              |1           |
+-----------------------------+------------+
|too-many-arguments           |1           |
+-----------------------------+------------+
|superfluous-parens           |1           |
+-----------------------------+------------+
|redefined-outer-name         |1           |
+-----------------------------+------------+
|no-self-use                  |1           |
+-----------------------------+------------+

```

The top on the list items are easy to fix. Also I think we should be flexible with line length. 80 characters seems kind of short nowadays with these large screens everywhere. I think it is a good length to aim for, but we should probably be flexible and allow for a bit more, say 90-100. 

The `too-many` are hard to deal with and I think we should disable those, some of the limits are pretty random anyways. A spectrum or the plotting class just need lots of methods and instances, and often it is hard to limit branches and statements.

The others are probably worth dealing with on case-by-case basis, I haven't checked the code for all of those.

Also we should think about if we use `pylint` on tests-fles too. Some of the tests cause errors in `pylint`, not just warnings, which is definitely an issue if we want to do that. `pyline` seems to not understand all the `pytest` syntax. Is there a special `pylint` instant or something for `pytest`?

**Comment by jccurtis** [comment]
The line length is a contentious issue like tabs v spaces (obviously spaces 😂). I have not had difficulty across many projects keeping my code to <80 characters. It also allows ease of coding side-by-side.

**Comment by cosama** [comment]
Just checked, it is in the [google style guide](https://github.com/google/styleguide/blob/gh-pages/pyguide.md), as are 4 spaces of intentions. These guys thought about that stuff way more than we ever will, so lets just stick to what they propose there IMHO.

**Comment by jccurtis** [comment]
Nice one @cosama! Google is echoing the general PEP8 guidelines. I'm all about it.

**Comment by jccurtis** [comment]
Thanks @cosama! We should be particularly careful to not create conflicts with the active feature branches that don’t have these changes. That said we should go through the feature branches next week and potentially prune stale ones.

**Comment by markbandstra** [comment]
Should we add `pylint` to our continuous integration configuration and expect the user to execute it on their own? E.g., the README would say for testing to run:
```
python setup.py test
pylint becquerel
```

**Comment by cosama** [comment]
I agree this might make merging in the future a bit more tricky, and we should carefully elaborate how to deal with this.

I think users should run `pylint` and `pytest` themselves locally before they submit a pull request and we should have instructions in the CONTRIBUTING.md which commands exactly to run. It should be also mentioned to run `pytest` for both python2 and python3.

Aside of this, I think adding it to continuous integration makes sense.

**Comment by cosama** [comment]
Is there a reason to choose google over pep8: https://www.python.org/dev/peps/pep-0008/

Pep8 is community supported while the google style guide just a company policy.

There is also a pep8 checker, so that would be easy to use instead of pylint.

**Comment by jccurtis** [comment]
I just stumbled upon this [open source google tool for formatting python](https://github.com/google/yapf). I just ran this with the default params on the entire repo. This defaults to `pep8` but we could also select `google` formatting style. Check out the just pushed `feature-yapf-formatting` branch. 😄

**Comment by markbandstra** [comment]
Good find. Any idea if it enforces Google style docstring formatting, which was one of the style points we were most interested in using?

**Comment by jccurtis** [comment]
I think we can use this for updating the docstrings: https://github.com/dadadel/pyment. It supports google styled docstrings.

**Comment by markbandstra** [comment]
Regarding instructions for CONTRIBUTING, what about these tests and checks:
```
python setup.py test
pycodestyle becquerel tests  # flake8 seems broken
pydocstyle becquerel tests
pylint becquerel tests
```
I added `pycodestyle` because `flake8` seems broken right now... `pycodestyle` is the new name for `pep8`.

**Comment by cosama** [comment]
I think that would work.

I wonder if we should add running `yapf` to the RELEASING.md checklist? Or would it make sense to do so all the time, or never?

**Comment by markbandstra** [comment]
I think we should suggest `yapf` and `pyment` (once we check them out, and/or some other tools) in CONTRIBUTING as possible ways to bring contributed code into style compliance. We should run them as part of the RELEASING process as well, but just as a check.

**Comment by jccurtis** [comment]
Another helpful tool to make docstring production easier: https://atom.io/packages/docblock-python

**Comment by jccurtis** [comment]
I think @markbandstra's tests above and advice in `CONTRIBUTING.md` to make compliant code by using the following:

- Appropriate code fixes for `pytest` failures
- `yapf` for code structure cleanup for `pycodestyle`/`pylint` failures
-  `pyment` and `docblock` in Atom for `pydocstyle` failures

**Comment by markbandstra** [comment]
- [ ] Add `yapf` and `pyment` commands to `CONTRIBUTING`
- [ ] Possibly have travis run them
- [ ] Add them to `requirements-dev.txt`
